cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i

```diff
@@ -607,64 +818,279 @@ void UnorderedRespHelper<algorithmFPType, cpu>::computeHistManyClasses(IndexType
     int idxFeatureBestSplit     = -1; //index of best feature value in the array of sorted feature values
-    for (size_t i = 0; i < nDiffFeatMax; ++i)
+
+    size_t minidx = 0;
+    size_t maxidx = nDiffFeatMax - 1;
+    size_t idx;
+
+    for (; (minidx < maxidx) && isZero<IndexType, cpu>(nFeatIdx[minidx]); minidx++)
+        ;
+
+    for (; (minidx < maxidx) && isZero<IndexType, cpu>(nFeatIdx[maxidx]); maxidx--)
+        ;
+
+    DAAL_ASSERT(minidx < maxidx); //if the if statement after minidx search doesn't activate, we have an issue.
+    if ((nFeatIdx[minidx] == n)   //last split
+        || ((n - nFeatIdx[minidx]) < nMinSplitPart) || ((totalWeights - featWeights[minidx]) < minWeightLeaf))
+        return idxFeatureBestSplit;
+
+    if (split.featureUnordered)
     {
-        if (!nFeatIdx[i]) continue;
-        algorithmFPType thisFeatWeights = featWeights[i];
+        RNGs<size_t, cpu> rng;
+        rng.uniform(1, &idx, this->engineImpl->getState(), minidx, maxidx); //find random index between minidx and maxidx
+        //iterate idx down for FinalizeBestSplit (since it splits leftward)
+        for (; (minidx < idx) && isZero<IndexType, cpu>(nFeatIdx[idx]); idx--)
+            ;
 
-        nLeft       = (split.featureUnordered ? nFeatIdx[i] : nLeft + nFeatIdx[i]);
-        leftWeights = (split.featureUnordered ? thisFeatWeights : leftWeights + thisFeatWeights);
-        if ((nLeft == n) //last split
-            || ((n - nLeft) < nMinSplitPart) || ((totalWeights - leftWeights) < minWeightLeaf))
-            break;
+        nLeft       = nFeatIdx[idx];
+        leftWeights = featWeights[idx];
 
-        if (!split.featureUnordered)
+        PRAGMA_IVDEP
+        PRAGMA_VECTOR_ALWAYS
+        //one against others
+        for (size_t iClass = 0; iClass < this->_nClasses; ++iClass) histLeft[iClass] = nSamplesPerClass[idx * this->_nClasses + iClass];
+    }
+    else
+    {
+        //randomly select a histogram split index
+        algorithmFPType fidx   = 0;
+        algorithmFPType minval = minidx ? this->indexedFeatures().min(iFeature) : this->indexedFeatures().binRightBorder(iFeature, minidx - 1);
+        algorithmFPType maxval = this->indexedFeatures().binRightBorder(iFeature, maxidx);
+        size_t mid;
+        size_t l = minidx;
+        idx      = maxidx;
+        RNGs<algorithmFPType, cpu> rng;
+        rng.uniform(1, &fidx, this->engineImpl->getState(), minval, maxval); //find random index between minidx and maxidx
+
+        while (l < idx)
         {
-            PRAGMA_IVDEP
-            PRAGMA_VECTOR_ALWAYS
-            for (size_t iClass = 0; iClass < _nClasses; ++iClass) histLeft[iClass] += nSamplesPerClass[i * _nClasses + iClass];
+            mid = l + (idx - l) / 2;
+            if (this->indexedFeatures().binRightBorder(iFeature, idx) > fidx)
+            {
+                idx = mid;
+            }
+            else
+            {
+                l = mid + 1;
+            }
         }
```

Comment:
This binary search code repeats at least couple of times in this file, and I think in 'regression' part also. Please consider moving into a separate method to reduce code duplication.

Reply:
Will do.